roles/translator/src/downstream_sv1/diff_management.rs

```diff
@@ -0,0 +1,354 @@
+        Downstream::init_difficulty_management(downstream.clone()).unwrap();
+        let mut target = initial_target;
+        let mut count = 0;
+        while elapsed <= total_run_time {
+            // start hashing util a target is met and submit to
+            mock_mine(target.clone().into());
+            Downstream::save_share(downstream.clone()).unwrap();
+            Downstream::try_update_difficulty_settings(downstream.clone())
+                .await
+                .unwrap();
+            target = downstream
+                .safe_lock(|d| {
+                    roles_logic_sv2::utils::hash_rate_to_target(
+                        d.difficulty_mgmt.min_individual_miner_hashrate,
+                        config_shares_per_minute,
+                    )
+                })
+                .unwrap();
+            elapsed = timer.elapsed();
+            count += 1;
+            println!("Submitted {:?} share in {:?} seconds", count, elapsed);
+            let calculated_share_per_min = count as f32 / (elapsed.as_secs_f32() / 60.0);
+            println!("Actual Share/Min {:?}", calculated_share_per_min);
+        }
+
+        let calculated_share_per_min = count as f32 / (elapsed.as_secs_f32() / 60.0);
+
+        println!(
+            "CALCULATED HASHRATE: {:?}",
+            downstream
+                .clone()
+                .safe_lock(|d| d.difficulty_mgmt.min_individual_miner_hashrate)
+                .unwrap()
+        );
+        println!("Actual Share/Min {:?}", calculated_share_per_min);
+        let calculated_share_per_min = count as f32 / (elapsed.as_secs_f32() / 60.0);
+        let err = 1.0
+            - (config_shares_per_minute - calculated_share_per_min / config_shares_per_minute)
+                .abs();
+        assert!(
+            err < 0.2,
+            "Calculated share_per_min does not meet 20% error"
+        );
+    }
+
+    fn mock_mine(target: Target) -> U256<'static> {
+        let mut share: Target = [255_u8; 32].into();
+        while shares_is_gt(share.clone().into(), target.clone().into()) {
+            let mut rng = thread_rng();
+            let number = rng.gen::<u64>();
+            let digest: U256 = Sha256::digest(&number.to_le_bytes())
+                .to_vec()
+                .try_into()
+                .unwrap();
+            share = u256_to_target(digest);
+        }
+        let test_share: U256 = share.clone().into();
+        let test_target: U256 = target.into();
+        println!("\nSHARE: {:?}", test_share);
+        println!("TARGET: {:?}\n", test_target);
+        share.into()
+    }
+
+    // returns hashrate based on how fast the device hashes over the given duration
+    fn measure_hashrate(duration_secs: u64) -> f64 {
+        let start_time = Instant::now();
+        let mut hashes: u64 = 0;
+        let duration = Duration::from_secs(duration_secs);
+
+        while start_time.elapsed() < duration {
+            let mut rng = thread_rng();
```

Comment:
This block is the same as above - could be a function - 
gen_rand_share() -> U256